driver/csiplugin/controllerserver.go

```diff
@@ -874,12 +894,167 @@ func (cs *ScaleControllerServer) ControllerPublishVolume(ctx context.Context, re
 	return &csi.ControllerPublishVolumeResponse{}, nil
 }
 
+//CreateSnapshot Create Snapshot
 func (cs *ScaleControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	glog.V(3).Infof("CreateSnapshot - create snapshot req: %v", req)
+
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		glog.V(3).Infof("CreateSnapshot - invalid create snapshot req: %v", req)
+		return nil, status.Error(codes.Internal, fmt.Sprintf("CreateSnapshot ValidateControllerServiceRequest failed: %v", err))
+	}
+
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "CreateSnapshot - Request cannot be empty")
+	}
+
+	volID := req.GetSourceVolumeId()
+	if volID == "" {
+		return nil, status.Error(codes.InvalidArgument, "CreateSnapshot - Source Volume ID is a required field")
+	}
+
+	volumeIDMembers, err := cs.GetVolIdMembers(volID)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("CreateSnapshot - Error in source Volume ID %v: %v", volID, err))
+	}
+
+	if !volumeIDMembers.IsFilesetBased {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("CreateSnapshot - Source Volume %v is not fileset based", volID))
+	}
+
+	conn, err := cs.GetConnFromClusterID(volumeIDMembers.ClusterId)
+	if err != nil {
+		return nil, err
+	}
+
+	filesystemName, err := conn.GetFilesystemName(volumeIDMembers.FsUUID)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("CreateSnapshot - Unable to get filesystem Name for Filesystem Uid [%v] and clusterId [%v]. Error [%v]", volumeIDMembers.FsUUID, volumeIDMembers.ClusterId, err))
+	}
+
+	glog.V(5).Infof("CreateSnapshot - getting filesystem Name from Filesystem Uid [%s] ", volumeIDMembers.FsetId)
+	filesetName, err := conn.GetFileSetNameFromId(filesystemName, volumeIDMembers.FsetId)
```

Comment:
Currently we do not have check for dependant vs independent fileset since we plan to support snapshot only for independent fileset

Reply:
Done